File: Array/67.二进制求和.java
// https://leetcode-cn.com/problems/add-binary/
// 官方题解 https://leetcode-cn.com/problems/add-binary/solution/er-jin-zhi-qiu-he-by-leetcode-solution/

class Solution {
    public String addBinary(String a, String b){
        // StringBuffer is powerful
        StringBuffer ans = new StringBuffer();

        int n = Math.max(a.length(), b.length()), carry = 0;
        for (int i=0; i < n; i++){
            carry += i < a.length() ? (a.charAt(a.length() - 1 -i) - '0') : 0;
            carry += i < b.length() ? (b.charAt(b.length() - 1 -i) - '0') : 0;
            ans.append((char)(carry % 2 + '0'));
            carry /= 2;
        }
        if (carry > 0){
            ans.append('1');
        }
        ans.reverse();
        return ans.toString();
    }
}

"""
class Solution {
    public String addBinary(String a, String b) {
        //自带的函数 (java无法通过全部测试用例)
        return Integer.toBinaryString(
            Integer.parseInt(a,2) + Integer.parseInt(b,2)
        );
        /*
        在 Java 中:
        如果字符串超过 3333 位，不能转化为 Integer
        如果字符串超过 6565 位，不能转化为 Long
        如果字符串超过 500000001500000001 位，不能转化为 BigInteger
        因此，为了适用于长度较大的字符串计算，我们应该使用更加健壮的算法。
        */
    }
}
"""

File: Array/67.二进制求和.py
# https://leetcode-cn.com/problems/add-binary/

class Solution:
    def addBinary(self, a: str, b: str) -> str:
        ## 二进制求和
        # 1.str --> int，并且值为二进制对应的值
        n1 = 0
        for i in range(len(a)):
            n1 = n1*2 + int(a[i])
        n2 = 0
        for i in range(len(b)):
            n2 = n2*2 + int(b[i])
        res = bin(n1 + n2)
        return str(res[2:])

# eval()函数用来执行一个字符串表达式，并返回表达式的值。
"""
class Solution:
    def addBinary(self, a: str, b: str) -> str:
        a = eval("0b"+a)
        b = eval("0b"+b)
        return bin(a+b)[2:]
"""

# 更简洁的解法 int(x, base) 第二个参数为基，默认十进制
"""
class Solution:
    def addBinary(self, a: str, b: str) -> str:
        return bin(int(a,2)+int(b,2))[2:]
"""